Logging updates

maradionov opened the pull request:
Add new log level METRIC (between INFO and WARNING) to print only metric information. Send metrics to tensorboard. Add example to PyTorch MNIST notebook.

maradionov committed:
Add new log level - METRIC. Write metrics on aggregator to tensorboard

maradionov committed:
Refactor code. Add tensorboardX to setup.py

maradionov committed:
fix flake8

itrushkin commented on openfl-workspace/default/plan/plan.yaml:
```diff
@@ -9,6 +9,7 @@ aggregator :
     best_state_path : save/best.pbuf
     last_state_path : save/last.pbuf
     rounds_to_train : 10
+    tensor_board    : false
```
```suggestion
    tensorboard    : false
```
We probably should preserve naming from [tensorflow/tensorboard](https://github.com/tensorflow/tensorboard) and [lanpa/tensorboardX](https://github.com/lanpa/tensorboardx).

alexey-gruzdev replied:
Agree with @itrushkin for naming

itrushkin commented on openfl-workspace/torch_cnn_mnist/plan/plan.yaml:
```diff
@@ -9,6 +9,7 @@ aggregator :
     best_state_path : save/torch_cnn_mnist_best.pbuf
     last_state_path : save/torch_cnn_mnist_last.pbuf
     rounds_to_train : 10
+    tensor_board    : true
```
```suggestion
    tensorboard    : true
```

itrushkin commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -103,6 +105,7 @@ def __init__(self,
         self.collaborator_tasks_results = {}
         # {TaskResultKey: data_size}
         self.collaborator_task_weight = {}
+        self.tb_writer = SummaryWriter(self.log_dir, flush_secs=10) if tensor_board else None
```
Could users pass `SummaryWriter` instance instead of the flag deciding the TensorBoard usage? It would allow users to define the logging directory, flush frequency, and other keyword arguments of `SummaryWriter`. Moreover, in this case, we do not need `tensorboardX` as a dependency.

maradionov replied:
resolve 747b5c53fee5da2143a617db0311e523c7c937d9

maradionov committed:
Return back global logger in native.py

maradionov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -41,6 +42,7 @@ def __init__(self,
                  assigner,
 
                  rounds_to_train=256,
+                 tensor_board=False,
```
```suggestion
                 tensorboard=False,
```

maradionov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -103,6 +105,7 @@ def __init__(self,
         self.collaborator_tasks_results = {}
         # {TaskResultKey: data_size}
         self.collaborator_task_weight = {}
+        self.tb_writer = SummaryWriter(self.log_dir, flush_secs=10) if tensor_board else None
```
```suggestion
        self.tb_writer = SummaryWriter(self.log_dir, flush_secs=10) if tensorboard else None
```

aleksandr-mokrov commented on openfl/utilities/utils.py:
```diff
@@ -3,6 +3,40 @@
+    if not methodName:
+        methodName = levelName.lower()
+
+    if hasattr(logging, levelName):
```
It looks redundant.  We don't catch this exception and I think that it is not a problem if we would try add the same level twiсe into logging. In this case, it is enough to exit.

maradionov replied:
done 3ed9f513adc09160ee04a63eb3e910b3b0fdfa65

maradionov committed:
Delete raising exceptions in addLoggingLevel function

maradionov committed:
Modify log level checking

maradionov committed:
Apply suggestions from code review

Rename tensorboard option

Co-authored-by: Ilya Trushkin <76161256+itrushkin@users.noreply.github.com>

maradionov committed:
Remove log level checking

dmitryagapov commented on openfl-workspace/default/plan/plan.yaml:
```diff
@@ -9,6 +9,7 @@ aggregator :
     best_state_path : save/best.pbuf
     last_state_path : save/last.pbuf
     rounds_to_train : 10
+    tensorboard    : false
```
align familiare with other strings

dmitryagapov commented on openfl-workspace/torch_cnn_mnist/plan/plan.yaml:
```diff
@@ -9,6 +9,7 @@ aggregator :
     best_state_path : save/torch_cnn_mnist_best.pbuf
     last_state_path : save/torch_cnn_mnist_last.pbuf
     rounds_to_train : 10
+    tensorboard    : true
```
align familiar with other strings

dmitryagapov commented on openfl/interface/cli.py:
```diff
@@ -7,35 +7,40 @@
 from pathlib import Path
+import os
 
+from openfl.utilities import addLoggingLevel
```
https://www.python.org/dev/peps/pep-0008/#imports
Imports should be grouped in the following order:

Standard library imports.
Related third party imports.
Local application/library specific imports.
You should put a blank line between each group of imports.

The same recommendation for all imports.

maradionov replied:
solved 79e053176690b2e97bcb12e9e5a66314d211dc02

dmitryagapov commented on openfl/interface/cli.py:
```diff
@@ -7,35 +7,40 @@
-        }
+    if isinstance(level, str):
+        level = level.upper()
 
-    if level.lower() in ['debug', 'error']:
-        os.environ['GRPC_VERBOSITY'] = level.upper()
+    if level in ['DEBUG', 'ERROR']:
+        os.environ['GRPC_VERBOSITY'] = level
```
When you use isinstance it looks like you want to handle both types of level `int` and `str`, as it used in logging module, but here only `str` expected and handeled on line 29.

maradionov replied:
solved 95fcf3916295687c77d7d34fc04074389b034e59

dmitryagapov commented on openfl/native/native.py:
```diff
@@ -39,12 +41,13 @@ def setup_plan(log_level='CRITICAL'):
     cols_config = 'plan/cols.yaml'
     data_config = 'plan/data.yaml'
 
+    curent_level = logging.root.level
```
*current :)

maradionov replied:
solved 79e053176690b2e97bcb12e9e5a66314d211dc02

dmitryagapov commented on openfl/utilities/utils.py:
```diff
@@ -3,6 +3,33 @@
+import logging
+
+
+def addLoggingLevel(levelName, levelNum, methodName=None):
```
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

maradionov replied:
solved 79e053176690b2e97bcb12e9e5a66314d211dc02

maradionov committed:
Pass write_metric function through plan

alexey-gruzdev commented:
@maradionov let's also check how it would work with new interactive API.

maradionov committed:
Merge branch 'develop' into logging_updates

maradionov committed:
Pass write metric function also in workspace

If we dont define this callback in notebook it will be taken from
workspace, if exists

maradionov committed:
Add docs for log metric callback

maradionov committed:
Clear notebook output

maradionov committed:
Merge branch 'develop' into logging_updates

maradionov committed:
Set 'metric' tag to train task result in interactive api

It's necessary for logger to parse this tag

maradionov committed:
Log metric task result on aggregator additionally.

Collaborator logs metric task result, but sometimes collaborator
and aggregator have different consoles (in interactive api e.g.)

maradionov committed:
fix flake8

maradionov committed:
Remove default tensorboard option in defaults plan.yaml

maradionov committed:
Remove tensorboardX dependency

maradionov committed:
Fix some pep8 issues. Remove debugging grpc code.

maradionov committed:
Remove debugging grpc code.

maradionov committed:
Merge branch 'develop' into logging_updates

maradionov committed:
fix flake8

maradionov committed:
Merge branch 'develop' into logging_updates

maradionov committed:
fix flake8

maradionov committed:
Add tensorboard to torch_cnn_mnist requirements

maradionov committed:
Fix typo

alexey-gruzdev commented:
Closed this since we've merged it to #104
